processing/src/main/java/org/apache/druid/segment/data/FrontCodedIndexed.java

```diff
@@ -0,0 +1,393 @@
+      {
+        return currIndex < adjustedNumValues;
+      }
+
+      @Override
+      public ByteBuffer next()
+      {
+        // null is handled special
+        if (hasNull && currIndex == 0) {
+          currIndex++;
+          return null;
+        }
+        if (!hasNext()) {
+          throw new NoSuchElementException();
+        }
+        final int adjustedCurrIndex = hasNull ? currIndex - 1 : currIndex;
+        final int bucketNum = adjustedCurrIndex >> div;
+        // load next bucket if needed
+        if (bucketNum != currentBucketIndex) {
+          final int offset = copy.getInt(offsetsPosition + ((bucketNum - 1) * Integer.BYTES));
+          copy.position(bucketsPosition + offset);
+          currentBucket = readBucket(
+              copy,
+              bucketNum < (numBuckets - 1) ? bucketSize : lastBucketNumValues
+          );
+          currentBucketIndex = bucketNum;
+        }
+        int offset = adjustedCurrIndex & rem;
+        currIndex++;
+        return currentBucket[offset];
+      }
+
+      @Override
+      public void remove()
+      {
+        throw new UnsupportedOperationException();
+      }
+    };
+  }
+
+  @Override
+  public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+  {
+    inspector.visit("buffer", buffer);
+    inspector.visit("hasNulls", hasNull);
+    inspector.visit("bucketSize", bucketSize);
+  }
+
+  /**
+   * Get a value from a bucket at a relative position
+   */
+  public static ByteBuffer getFromBucket(ByteBuffer bucket, int offset)
+  {
+    final byte[] first = readBytes(bucket);
+    if (offset == 0) {
+      return ByteBuffer.wrap(first);
+    }
+    int pos = 0;
+    int prefixLength;
+    final byte[] fragment;
+    // scan through bucket values until we reach offset
+    do {
+      prefixLength = VByte.readInt(bucket);
+      if (++pos < offset) {
+        final int skipLength = VByte.readInt(bucket);
+        bucket.position(bucket.position() + skipLength);
+      } else {
+        fragment = readBytes(bucket);
+        break;
+      }
+    } while (true);
+    return combinePrefixAndFragment(first, prefixLength, fragment);
+  }
+
+  /**
+   * Find the relative position of a value in a bucket with a linear scan
+   */
+  public static int findInBucket(ByteBuffer bucket, int numValues, ByteBuffer value, Comparator<ByteBuffer> comparator)
+  {
+    final byte[] first = readBytes(bucket);
+    final int firstCompare = comparator.compare(ByteBuffer.wrap(first), value);
+    if (firstCompare == 0) {
+      return 0;
+    }
+    if (firstCompare > 0) {
+      return -1;
+    }
+    int offset = 0;
+    int prefixLength;
+    byte[] fragment;
+    // scan through bucket values until we find match or compare numValues
+    int insertionPoint = 1;
+    while (++offset < numValues) {
+      prefixLength = VByte.readInt(bucket);
+      fragment = readBytes(bucket);
+      ByteBuffer next = combinePrefixAndFragment(first, prefixLength, fragment);
```

Comment:
combining all the time to then compare seems quite wasteful, can we not push this comparison down to the buffer instead of copying it out like this?

For example, given that we know the target value, we should know the size of prefix that matches it by this point.  We should be able to read the prefix length and make sure they are the same, if they aren't the same, then no match, move on.  Though, that works if we only care about equality.  If we want to also have mathematical comparisons, we likely need to be a bit more intelligent, but can still use known prefixes and other things pushed all the way down to the buffer instead of the naive read, build buffer, compare via generic comparator approach here.

Reply:
I guess this would be an additional restriction on use that the current implementation doesn't have, which is that it requires the prefix and suffix values be complete values of their own.

This does work for strings though, and is probably worth doing to save the cycles since its a hot spot, but I wonder if its overly restrictive to prevent re-use for other types of dictionary encoded columns. Or maybe we should just specialize this for utf8 strings only?

Reply:
i think the optimizing the comparison operations require dropping the comparator and specializing this implementation to be used strictly for things that are ok with unsigned byte comparison, which I'm ok with, but does limit the potential to re-use with other types